File: src/my_project/medium_problems/from51to100/search_a_2d_matrix_log.py
from typing import List 

class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        nums = []
        
        for list_nums in matrix:
            nums = nums + list_nums
        
        def binarySearch(nums: List[int], left: int, right: int, target:int) -> bool:
            if left > right:
                return False 

            mid = (left + right) // 2 

            if nums[mid] == target:
                return True

            if nums[left] <= nums[mid]:

                if target >= nums[left] and target <= nums[mid]:
                    return binarySearch(nums, left, mid-1,target)
                return binarySearch(nums, mid + 1, right, target)

            if target >= nums[mid] and target <= nums[right]:
                return binarySearch(nums, mid + 1, right, target)
            return binarySearch(nums, left, mid -1,target)


        return binarySearch(nums, 0, len(nums)-1, target)

File: my_project_typescript/medium_problems_typescript/from1to50/searchA2dMatrixLog.ts
function searchMatrix(matrix: number[][], target: number): boolean {

    let nums: Array<number> = []

    for (let arrayNum of matrix){
        nums = nums.concat(arrayNum)
    }

    function binarySearch(nums: number[], left: number, right: number, target:number):boolean{

        if (left > right){
            return false
        }

        let mid: number = Math.floor((left+right)/2)

        if (nums[mid] === target){
            return true
        }

        if (nums[left] <= nums[mid]){
            if (target>=nums[left] && target<=nums[mid]){
                return binarySearch(nums, left, mid-1, target)
            }
            return binarySearch(nums, mid+1,right, target)
        }

        if (target>=nums[mid] && target<=nums[right]){
            return binarySearch(nums, mid+1, right, target)
        }
        return binarySearch(nums, left, mid-1, target)
    }

    return binarySearch(nums, 0, nums.length-1, target)

};